core/src/main/scala/kafka/server/metadata/ZkMetadataCache.scala

```diff
@@ -55,6 +56,64 @@ trait ZkFinalizedFeatureCache {
   def getFeatureOption: Option[Features]
 }
 
+case class MetadataSnapshot(partitionStates: mutable.AnyRefMap[String, mutable.LongMap[UpdateMetadataPartitionState]],
+                            topicIds: Map[String, Uuid],
+                            controllerId: Option[CachedControllerId],
+                            aliveBrokers: mutable.LongMap[Broker],
+                            aliveNodes: mutable.LongMap[collection.Map[ListenerName, Node]]) {
+  val topicNames: Map[Uuid, String] = topicIds.map { case (topicName, topicId) => (topicId, topicName) }
+}
+
+object ZkMetadataCache {
+  /**
+   * Create topic deletions (leader=-2) for topics that are missing in a FULL UpdateMetadataRequest coming from a
+   * KRaft controller during a ZK migration.
+   */
+  def maybeInjectDeletedPartitions(
+    currentMetadata: MetadataSnapshot,
+    updateMetadataRequest: UpdateMetadataRequest
+  ): Seq[Uuid] = {
+    if (updateMetadataRequest.isKRaftController && updateMetadataRequest.updateType() == AbstractControlRequest.Type.FULL) {
```

Comment:
This seems to duplicate the check below, where we only invoke this function if `updateType == FULL`.

If we are going to check this here (not sure if we even should) seems like we should just throw an exception if `updateType != FULL` here?

Reply:
Yea, we can do without this check here.